lxd/identities.go

```diff
@@ -277,137 +350,198 @@ func identityAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *http
 		}
 
-		// Filter results by what the user is allowed to view.
-		for _, id := range allIdentities {
-			if canView(id) {
-				identities = append(identities, id)
+		var identities []dbCluster.Identity
+		var groupsByIdentityID map[int][]dbCluster.AuthGroup
+		var apiIdentity *api.Identity
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			// Get all identities, filter by authentication method if present.
+			var filters []dbCluster.IdentityFilter
+			if authenticationMethod != "" {
+				clusterAuthMethod := dbCluster.AuthMethod(authenticationMethod)
+				filters = append(filters, dbCluster.IdentityFilter{AuthMethod: &clusterAuthMethod})
 			}
-		}
-
-		if len(identities) == 0 {
-			return nil
-		}
 
-		if recursion == "1" && len(identities) == 1 {
-			// It's likely that the user can only view themselves. If so we can optimise here by only getting the
-			// groups for that user.
-			apiIdentity, err = identities[0].ToAPI(ctx, tx.Tx(), canViewGroup)
+			allIdentities, err := dbCluster.GetIdentitys(ctx, tx.Tx(), filters...)
 			if err != nil {
 				return err
 			}
-		} else if recursion == "1" {
-			// Otherwise, get all groups and populate the identities outside of the transaction.
-			groupsByIdentityID, err = dbCluster.GetAllAuthGroupsByIdentityIDs(ctx, tx.Tx())
-			if err != nil {
-				return err
+
+			// Filter results by what the user is allowed to view.
+			for _, id := range allIdentities {
+				if canView(id) {
+					identities = append(identities, id)
+				}
 			}
-		}
 
-		return nil
-	})
-	if err != nil {
-		return response.SmartError(err)
-	}
+			if len(identities) == 0 {
+				return nil
+			}
 
-	// Optimisation for user that can only view themselves.
-	if apiIdentity != nil {
-		return response.SyncResponse(true, []api.Identity{*apiIdentity})
-	}
+			if recursion == "1" && len(identities) == 1 {
+				// It's likely that the user can only view themselves. If so we can optimise here by only getting the
+				// groups for that user.
+				apiIdentity, err = identities[0].ToAPI(ctx, tx.Tx(), canViewGroup)
+				if err != nil {
+					return err
+				}
+			} else if recursion == "1" {
+				// Otherwise, get all groups and populate the identities outside of the transaction.
+				groupsByIdentityID, err = dbCluster.GetAllAuthGroupsByIdentityIDs(ctx, tx.Tx())
+				if err != nil {
+					return err
+				}
+			}
 
-	if recursion == "1" {
-		// Convert the []cluster.Group in the groupsByIdentityID map to string slices of the group names.
-		groupNamesByIdentityID := make(map[int][]string, len(groupsByIdentityID))
-		for identityID, groups := range groupsByIdentityID {
-			for _, group := range groups {
-				if canViewGroup(entity.AuthGroupURL(group.Name)) {
-					groupNamesByIdentityID[identityID] = append(groupNamesByIdentityID[identityID], group.Name)
+			return nil
+		})
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// Optimisation for user that can only view themselves.
```

Comment:
adding a *why* here would be useful (why does `apiIdentity != nil ` mean they can only view themselves)?

Reply:
This hasn't changed from before but I'll add a commit to clarify this :+1:

Reply:
Done.